lib/wadler/print.rb

```diff
@@ -0,0 +1,89 @@
+# frozen_string_literal: true
+
+# Oppen.
+module Oppen
+  # Wadler.
+  class Wadler
+    attr_reader :current_indent
+    attr_reader :margin
+    attr_reader :new_line
+    attr_reader :tokens
+
+    def initialize(margin: 80, new_line: "\n")
+      @current_indent = 0
+      @margin = margin
+      @new_line = new_line
+      @tokens = []
+    end
+
+    # @return [String]
+    def output
+      tokens << Oppen.eof
+      Oppen.print(tokens:, margin:, new_line:)
+    end
+
+    # @param indent [Integer] group indentation
+    # @param open_obj [String] group oppening delimiter
+    # @param close_obj [String] group closing delimiter
+    # @param break_type [Oppen::Token::BreakType] group breaking type
+    #
+    # @return [Nil]
+    def group(indent = 0, open_obj = '', close_obj = '',
+              break_type = Oppen::Token::BreakType::CONSISTENT, &)
+      if !open_obj.empty?
+        self.break
+        text(open_obj)
+      end
+
+      nest(indent, break_type, &)
+
+      if !close_obj.empty? # rubocop:disable Style/GuardClause
+        self.break
+        text(close_obj)
+      end
+    end
+
+    # @param indent [Integer] nest indentation
+    # @param break_type [Oppen::Token::BreakType] nest breaking type
+    #
+    # @return [Nil]
+    def nest(indent = 2, break_type = Oppen::Token::BreakType::CONSISTENT, &block)
+      previous_indent = current_indent
+      @current_indent = indent
+
+      tokens <<
+        case break_type
+        in Oppen::Token::BreakType::CONSISTENT
+          Oppen.begin_consistent(offset: indent)
+        in Oppen::Token::BreakType::INCONSISTENT
+          Oppen.begin_inconsistent(offset: indent)
+        end
+
+      block.call
+
+      tokens << Oppen.end
+      @current_indent = previous_indent
+    end
```

Comment:
`indent` for `nest` [in `prettyprint`](https://github.com/ruby/prettyprint/blob/ce5792bc81bd5d102902ddc0205f48e326f3c73c/lib/prettyprint.rb#L279) does not take an optional value.

I can understand `break_type`, because we introduced it, but we really need be sticking to the the `prettyprint` API.

I will add it to the tickets though. We need to merge this.